### PR TITLE
Servlet Filter for Authorization Header Validation

### PR DESCRIPTION
**Pull request by emswbandara**
### Proposed changes in this pull request

This PR provides a solution to the issue where an NPE is thrown by CXF transport exposing the stack-trace and server information to the end user for broken authorization headers. AuthorizationHeaderValidationServletFilter validates the authorization headers before it reaches CXF level and sends a proper error response back to the user for invalid header formats.

### When should this PR be merged

Immediately

### Follow up actions

None

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+
+/**
+ * Servlet Filter implementation class Authorization Header Filter
```
Comment (rushmin) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java]: Class comment should explain what the class does.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        this.context.log("AuthorizationHeaderFilter initialized");
+    }
+
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+
+        final HttpServletRequest req = (HttpServletRequest) request;
+        final String authHeader = req.getHeader(HTTPConstants.HEADER_AUTHORIZATION);
+        if (StringUtils.isEmpty(authHeader)) {
+            chain.doFilter(request, response);
```
Comment (rushmin) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java]: The method should return if we continue with other filters in the chain.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if (AUTH_TYPE_BASIC.equals(authType)) {
+            final String authCredentials = (authHeader.trim().substring(AUTH_TYPE_BASIC_LENGTH)).trim();
+            //if auth header comes in invalid format send error in response
+            if (StringUtils.isBlank(authCredentials) || authCredentials.indexOf(' ') >= 0
+                    || authCredentials.length() < MINIMUM_CREDENTIAL_SIZE) {
+                String errorMsg = "Internal Server Error";
+                handleErrorResponse((HttpServletResponse) response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+                        errorMsg);
+            }
+            //if there is more than single space between auth_type and credentials modify the request header
```
Comment (harsha1979) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java]: add this comment to inside the else condition

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java (hunk 2 above)
Comment (emswbandara) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/filter/AuthorizationHeaderFilter.java]: Fixed with  d152ebd

**Pull request merged**